src/hooks/useFetch.js

```diff
@@ -0,0 +1,51 @@
+import { useCallback, useEffect, useState } from 'react';
+
+export default function useFetch({
+  asyncFunction,
+  deps = [],
+  immediate = false,
+}) {
+  const [state, setState] = useState({
+    data: null,
+    loading: false,
+    error: null,
+  });
+  const refetch = useCallback(async () => {
+    setState((prev) => ({ ...prev, error: null, loading: true }));
+    try {
+      const response = await asyncFunction();
+      setState((prev) => ({ ...prev, loading: false, data: response }));
+    } catch (error) {
+      setState({ data: null, loading: false, error });
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [...deps]);
+
+  useEffect(() => {
+    if (immediate) refetch();
+  }, [immediate, refetch]);
```

Comment:
```suggestion
  const asyncFnRef = useRef(asyncFunction);
  useEffect(() => {
    asyncFnRef.current = asyncFunction;
  }, [asyncFunction]);

  const refetchRef = useRef(null);
  if (!refetchRef.current) {
    refetchRef.current = async () => {
      setState((prev) => ({ ...prev, error: null, loading: true }));
      try {
        const response = await asyncFnRef.current();
        setState((prev) => ({ ...prev, loading: false, data: response }));
      } catch (error) {
        setState({ data: null, loading: false, error });
      }
    };
  }

  useEffect(() => {
    if (immediate) refetchRef.current();
  }, [immediate, ...deps]);
```

이렇게 Ref 패턴을 사용하면 useCallback 훅을 사용하지 않아도 정상 동작해요. React Query에서 실제로 사용하는 방식이기도 합니다. 코드를 천천히 읽고, 분석해 보세요. 🤓

Reply:
제안 감사드립니다☺️!
알려주신 내용을 검색하여 몇개의 글을 읽어보았습니다. 좀 어려운 내용들인 것 같네요🥲. 그래서 추가적으로 수정을 하고 있는데, 두가지의 공부 확인겸 질문이 있습니다ㅎㅎㅎ

**1.**
[useRef에 대한 React 문서](https://ko.react.dev/reference/react/useRef#referencing-a-value-with-a-ref)를 읽어보니, `ref.current`를 할당할때 렌더링중 한번만 할당하기 위해서 useEffect를 사용하거나, null check 패턴을 사용하는 것 같습니다. 그렇다면 제안 해주신 코드도 일관성을 위해서 다음과 같이 null check로 통일해도 무방할까요? 동작은 동일한 것 같습니다.

```jsx
  const asyncFnRef = useRef(null);
  if (asyncFnRef.current === null) {
    asyncFnRef.current = asyncFunction;
  }

  const refetchRef = useRef(null);
  if (refetchRef.current === null) {
    refetchRef.current = async () => {
      setState((prev) => ({ ...prev, error: null, loading: true }));
      try {
        const response = await asyncFnRef.current();
        setState((prev) => ({ ...prev, loading: false, data: response }));
      } catch (error) {
        setState({ data: null, loading: false, error });
      }
    };
  }
```
**2.**
알고보니 Tanstack Query v5에서는 `onSuccess, onError`등을 오용 케이스 때문에 deprecated 했다고 하는데요. [해당 글](https://tkdodo.eu/blog/breaking-react-querys-api-on-purpose#the-callbacks-might-not-run)에서 Data Fetching 로직과 Side Effect 로직을 분리하는 코드를 권장하는 내용을 보았습니다. 같은 이유로 현재 코드의 `deps` 부분도 분리하는게 좋을까요? 

사실은 이 코드에서 오용될 여지는 잘 모르겠습니다. 다만, deps에서 `react-hooks/exhaustive-deps` 경고를 보니, useFetch의 호출부와 useEffect 사이 코드의 거리가 멀고, 여러 상태들을 deps에 넣다 보면, 무한 렌더링에 빠지기 쉬울 수 있을 것 같기도 합니다.


```jsx
 useEffect(() => {
    if (immediate) refetchRef.current();
  }, [immediate, ...deps]); 

// 현재 useFetch는 deps 받아서 refetch를 실행하고 있지만 해당 매개변수를 제거하고, 
// 호출부에서 다른 custom hook 혹은 useEffect로 Side Effect를 처리하기
```

Reply:
어려운 내용이지만, 같이 대화해 가면서 알아보죠! 😊

1.
수형님이 제안하신 코드라면, `asyncFnRef`는 만들지 않아도 되지 않을까요? 그럼 결국 아래의 코드와 같은 동작이 되고, 오래된 클로저(stale closure)가 생길 수 있어요.

```
 const refetchRef = useRef(null);
  if (refetchRef.current === null) {
    refetchRef.current = async () => {
      setState((prev) => ({ ...prev, error: null, loading: true }));
      try {
        const response = await asyncFunction();
        setState((prev) => ({ ...prev, loading: false, data: response }));
      } catch (error) {
        setState({ data: null, loading: false, error });
      }
    };
  }
```

그럼 결국, deps가 바뀌어도 최신 응답을 받을 수 없게 돼요. 수형님이 제안하신 코드로 페이지네이션을 테스트 해보면 알 수 있어요. 🙂 (useEffect를 사용한 건, `asyncFunction`의 응답을 최신값으로 업데이트 하기 위해 effect 안에 넣은 거예요.)

2.
맞아요! V5 버전부터 `useQuery`에 관련된 콜백들이 제거됐죠. `useRef`를 통해 `refetch`의 재생성을 막았기 때문에, deps가 많아져도 무한 렌더링은 되지 않을 거라고 생각해요. 다만, 이제 수형님이 `onSuccess`와 같은 기능을 추가하기 보다는 글에서 나온 것처럼 effect나 파생 상태로 관리하는 것을 제안드렸던 거예요. 🤓

Reply:
생각지도 못하게 stale closure와 closure에 대해서도 공부하게 되었네요ㅋㅋㅋ 감사합니다! (이래서 자바스크립트를 공부해야하는군요 😷)